lib/phoenix/live_dashboard/components/chart_component.ex

```diff
@@ -1,131 +1,76 @@
 defmodule Phoenix.LiveDashboard.ChartComponent do
   use Phoenix.LiveDashboard.Web, :live_component
 
-  @default_prune_threshold 1_000
-
-  @default_bucket_size 20
-
   @impl true
   def mount(socket) do
-    {:ok, socket, temporary_assigns: [data: []]}
+    {:ok, stream(socket, :data, [], dom_id: &data_dom_id/1)}
   end
 
+  defp data_dom_id(_), do: "unused"
+
   @impl true
   def update(assigns, socket) do
-    {metric, assigns} = Map.pop(assigns, :metric)
+    {data, assigns} = Map.pop(assigns, :data, [])
+    socket = assign(socket, assigns) |> normalize_assigns!(data)
+    {:ok, socket}
+  end
 
-    socket =
-      if metric do
-        kind = chart_kind(metric.__struct__)
+  defp normalize_assigns!(socket, data) do
+    %{assigns: assigns} = socket
+    validate_positive_integer_or_nil!(assigns[:bucket_size], :bucket_size)
+    validate_positive_integer_or_nil!(assigns[:prune_threshold], :prune_threshold)
+    normalize_data(socket, data)
+  end
+
+  defp validate_positive_integer_or_nil!(nil, _field), do: nil
+
+  defp validate_positive_integer_or_nil!(value, field) do
+    unless is_integer(value) and value > 0 do
+      msg = "#{inspect(field)} must be a positive integer, got: #{inspect(value)}"
+      raise ArgumentError, msg
+    end
+  end
 
-        socket
-        |> apply_metric(kind, metric)
-        |> assign(
-          title: chart_title(metric),
-          description: metric.description,
-          kind: kind,
-          label: chart_label(metric),
-          tags: Enum.join(metric.tags, "-"),
-          unit: chart_unit(metric.unit),
-          prune_threshold: prune_threshold(metric)
-        )
-      else
-        socket
-      end
+  defp normalize_data(socket, data) do
+    label = socket.assigns.label
 
-    {:ok, assign(socket, assigns)}
+    data
+    |> Enum.map(fn {x, y, z} -> {x || label, y, z} end)
+    |> Enum.reduce(socket, fn elem, socket -> stream_insert(socket, :data, elem) end)
   end
 
   @impl true
   def render(assigns) do
     ~H"""
-    <div class="col-xl-6 col-xxl-4 col-xxxl-3 charts-col">
+    <div class={chart_size(@full_width)}>
       <div id={"chart-#{@id}"} class="card">
         <div class="card-body">
+          <!-- We don't add a phx-update="stream" because js expects only new data -->
           <div phx-hook="PhxChartComponent" id={"chart-#{@id}-datasets"} hidden>
-          <%= for {x, y, z} <- @data do %>
-            <span data-x={x || @label} data-y={y} data-z={z}></span>
-          <% end %>
+            <span :for={{_id, {x, y, z}} <- @streams.data} data-x={x} data-y={y} data-z={z}></span>
```

Comment:
I added a comment because it seems very strange behaviour to me :)

Reply:
Right, the charts work "outside of LiveView", so streams don't really help much. Like we can't use them to delete or insert at given positions. We can use it as a way to collect data but then it works pretty much the same as temporary assigns.

Both approaches are fine though, your call.

Reply:
I changed to stream because I thought (and maybe I'm wrong) that temporary assigns would be deprecated in the future in favor of streams. I'll keep the stream version.